adds migration guide and options in handle_build (#938)
nvidia-holoscan/holohub: utilities/cli/README.md

# HoloHub CLI Migration Guide

## Overview

HoloHub has been refactored from bash-based CLI scripts (`./run` and `./dev_container`) to a modern, unified Python-based CLI tool (`./holohub`). This upgrade provides a cleaner, more powerful, and more intuitive development experience.

## Key Benefits of the Refactoring

### **Simplified Workflows**
Transform complex multi-command workflows into single, intuitive commands:

```bash
# OLD: Container development
./dev_container build_and_run myapp --language cpp

# NEW: Container development
./holohub run myapp --language cpp
```

```bash
# OLD: Local development
./run build myapp --type debug
./run launch myapp cpp

# NEW: Local development (single command)
./holohub run myapp --language cpp --local
```

### **Enhanced Developer Experience**

> - **Unified Interface**: One tool instead of juggling `./run` and `./dev_container`

Reviewer: ```suggestion
- **Unified Interface**: One tool for both local and containerized development instead of juggling `./run` and `./dev_container`
```

### **Quick Command Reference**
| Workflow | Old Commands | New Command |
|----------|-------------|-------------|
| **Container Build & Run** | `./dev_container build_and_run` | `./holohub run` |
| **Local Build & Run** | `./run build` + `./run launch` | `./holohub run --local` |
| **Container Build Only** | `./dev_container build` | `./holohub build-container` |
| **Local Build Only** | `./run build` | `./holohub build --local` |
| **Run Without Building** | `./run launch` | `./holohub run --local --no-local-build` |
| **Container Run Only** | `./dev_container launch` | `./holohub run-container --no-docker-build` |

## Migration Examples

### **Build & Run**
```bash
# Build locally with optional operators and build type
# previous ./run build myapp ...
./holohub build myapp --local --build-type debug --build-with "op1;op2"

# Build in container
# previous ./dev_container build ...
./holohub build myapp --build-type debug --verbose

# Run locally
# previous ./run build myapp && ./run launch myapp cpp
./holohub run myapp --local --language cpp --build-type debug

# Run in container
# previous ./dev_container build_and_run ...

> ./holohub run myapp --language=cpp --run-args="--config=config.json" # NEW

Reviewer: Why "# NEW"? Wasn't args forwarding supported in `dev_container`?

```suggestion
./holohub run myapp --language=cpp --run-args="--config=config.json"
```

# Install packages
# previous ./run build myapp --install
./holohub install myapp
```

### **Container Operations**
```bash
# Container build
# previous ./dev_container build ...
./holohub build-container --base-img nvidia/holoscan:latest

# Container run
# previous ./dev_container launch ...
./holohub run-container --img myapp:dev --ssh-x11 --docker-opts="--entrypoint=bash" --no-docker-build

# Development environment
# previous ./dev_container vscode ...
./holohub vscode myapp --language cpp
```

> ### **Advanced Features**

Reviewer: ```suggestion
### **Advanced Features**

Use `-h` or `--help` to view a complete list of commands or subcommand options.

```

```bash
# Custom build options
./holohub build myapp --build-with "op1;op2" --configure-args='-DCUSTOM=ON'

# Benchmarking an application (only for applications/workflows)
./holohub build myapp --benchmark
```

###  **Development Environment**

Development commands remain familiar:

```bash
# Environment setup (unchanged)
[sudo] ./holohub setup

# Enhanced linting with integrated dependency management
./holohub lint --install-dependencies  # Previously ./run install_lint_deps
./holohub lint --fix                   # Same clean interface

# Consistent naming conventions
./holohub clear-cache                  # Previously ./run clear_cache

> ./holohub list                         # Unchanged

Reviewer: ```suggestion
./holohub list                         # Previously ./run list
```

```

## New Capabilities & Enhancements

### **Project Creation and Testing**
The new CLI introduces development lifecycle features:

```bash
# Scaffold new projects with templates
./holohub create my_new_app --language cpp

# Integrated testing framework
./holohub test myapp

# Show environment information
./holohub env-info
```

> ### **Intelligent Development Experience**
> - **Smart Error Messages**: Automatic project name suggestions for typos and better validation
> - **Modern Argument Parsing**: Consistent interface across all commands with improved help text
> - **Enhanced Container Integration**: Better argument forwarding and container lifecycle management

Reviewer: Seems redundant with introductory section?

```suggestion
```

### **Granular Build Control**

> The new architecture provides precise control over your development workflow:
> - **`--local`**: Explicit local development mode
> - **`--no-local-build`**: Skip application rebuild for faster iteration
> - **`--no-docker-build`**: Use existing container images
> - **Dedicated Commands**: Separate `build`, `run`, `build-container`, and `run-container` commands for clear workflow control

Author: ```suggestion
The new architecture provides precise control over your development workflow.

**Default Behavior:**
By default, `./holohub run` operates in a 'containerized mode', which means it will:
1. Build the container image (unless skipped, e.g. using `--no-docker-build`)
2. Build the application inside the container
3. Run the application inside the container

This container-first approach ensures consistency and reproducibility across different development environments.

**Command-line Options:**
- **`--local`**: Explicit local development mode
- **`--no-local-build`**: Skip application rebuild for faster iteration
- **`--no-docker-build`**: Use existing container images
- **Dedicated Commands**: Separate `build`, `run`, `build-container`, and `run-container` commands for clear workflow control

**Environment Variables:**
- **`HOLOHUB_BUILD_LOCAL`**: When set to any value, forces local mode (equivalent to always passing `--local`)
- **`HOLOHUB_ALWAYS_BUILD`**: Controls whether builds should be executed (defaults to `true`)
  - Set to `false` to skip both local and container builds
  - Useful for development iterations where you only want to run existing builds

```

> - **Dedicated Commands**: Separate `build`, `run`, `build-container`, and `run-container` commands for clear workflow control

Reviewer: Could we also add discussion here or in a followup section on:
1. Differences in default behavior (i.e. `./holohub run` is containerized by default and runs all containerized build/run steps)
2. Environment variables observed by the CLI that can alter default behavior (local build, ignore all previous build stages)

## Getting Help

```bash
./holohub --help              # General help
./holohub run --help          # Command-specific help
./holohub run myapp --verbose --dryrun  # Debug mode
./holohub list                # Check available projects
```

## Useful Commands

> - When adding option that looks like an argument, use `=` instead of whitespace ` `:

Reviewer: Could we add a note on why this is required? Is this the result of the argparse + shlex + docker entrypoint that we are using in the CLI?

```bash
--run-args="--verbose"   # instead of --run-args "--verbose"
```

> - To clear unused docker cache during development:
> ```
> docker image prune -f && docker buildx prune --filter type=regular -f
> ```

Reviewer: Suggest referencing Docker docs in case recommended approach changes in the future:
```suggestion
- To clear unused docker cache during development:
  - Clear Docker images ([doc](https://docs.docker.com/reference/cli/docker/image/prune/)): `docker image prune`
  - Clear Docker buildx cache ([doc](https://docs.docker.com/reference/cli/docker/buildx/prune/)): `docker buildx prune`
  - Clear Docker containers, networks, images ([doc](https://docs.docker.com/reference/cli/docker/system/prune/): `docker system prune`
````

Author: thanks, I was trying to document this specific command, but I think providing the generic pruning docs is also ok

# HoloHub CLI Migration Guide

## Overview

As of June 2025, HoloHub has been refactored from bash-based CLI scripts (`./run` and `./dev_container`) to a modern, unified Python-based CLI tool (`./holohub`). This upgrade provides a cleaner, more powerful, and more intuitive development experience.

## Key Benefits of the Refactoring

### **Simplified Workflows**
Transform complex multi-command workflows into single, intuitive commands:

```bash
# OLD: Container development
./dev_container build_and_run myapp --language cpp

# NEW: Container development
./holohub run myapp --language cpp
```

```bash

> # OLD: Local development
> ./run build myapp --type debug
> ./run launch myapp cpp
> 
> # NEW: Local development (single command)
> ./holohub run myapp --language cpp --local

Reviewer: Should we recommend `=` with `--language`? Also adding `--build-type` for parity in the example:
```suggestion
# OLD: Local development
./run build myapp --type debug
./run launch myapp cpp

# NEW: Local development (single command)
./holohub run myapp --language=cpp --local --build-type=debug
```

```

### **Enhanced Developer Experience**
- **Unified Interface**: One tool for both local and containerized development instead of juggling `./run` and `./dev_container`
- **Intelligent Defaults**: Container-first development for consistency and reproducibility
- **Better Error Messages**: Helpful suggestions when commands or project names don't match
- **Modern Python Implementation**: More reliable, maintainable, and extensible

### **Quick Command Reference**
| Workflow | Old Commands | New Command |
|----------|-------------|-------------|
| **Container Build & Run** | `./dev_container build_and_run` | `./holohub run` |
| **Local Build & Run** | `./run build` + `./run launch` | `./holohub run --local` |
| **Container Build Only** | `./dev_container build` | `./holohub build-container` |
| **Local Build Only** | `./run build` | `./holohub build --local` |
| **Run Without Building** | `./run launch` | `./holohub run --local --no-local-build` |
| **Container Run Only** | `./dev_container launch` | `./holohub run-container --no-docker-build` |

## Migration Examples

### **Build & Run**

> ```bash
> # Build locally with optional operators and build type
> # previous ./run build myapp ...
> ./holohub build myapp --local --build-type debug --build-with "op1;op2"

Reviewer: Suggest exact match for args between prev and new examples
```suggestion
```bash
# Build locally with optional operators and build type
# previous ./run build myapp --type debug --with "op1;op2"
./holohub build myapp --local --build-type debug --build-with "op1;op2"
```

> # Build in container
> # previous ./dev_container build ...
> ./holohub build myapp --build-type debug --verbose

Reviewer: Update for parity (`./dev_container build` is incomplete):

```suggestion
# Build in container
# Previous:
# $ ./dev_container build --docker_file path/to/myapp/dockerfile --img holohub:myapp
# $ ./dev_container launch --img holohub:myapp
# >>> ./run build myapp --type debug
# Updated:
./holohub build myapp --build-type debug
```

# Run locally
# previous ./run build myapp && ./run launch myapp cpp
./holohub run myapp --local --language cpp --build-type debug

> # Run in container
> # previous ./dev_container build_and_run ...
> ./holohub run myapp --language=cpp --run-args="--config=config.json"

Reviewer: ```suggestion
# Run in container
# previous ./dev_container build_and_run myapp cpp --run_args "--config=config.json"
./holohub run myapp --language=cpp --run-args="--config=config.json"
```

> # Install packages

Reviewer: Technically distinct from packaging operations (no deb output)

```suggestion
# Install built files
```

# previous ./dev_container build_and_install myapp
./holohub install myapp
```

### **Container Operations**
```bash
# Container build

> # previous ./dev_container build ...

Reviewer: ```suggestion
# previous ./dev_container build --docker_file <path/to/myapp/Dockerfile> --img holohub:myapp
```

./holohub build-container <my_project> [options]

# Container run

> # previous ./dev_container launch ...

Reviewer: ```suggestion
# previous ./dev_container launch --img holohub:myapp
```

./holohub run-container <my_project> --no-docker-build [options]

# Development environment

> # previous ./dev_container vscode ...

Reviewer: ```suggestion
# previous ./dev_container vscode myapp
```
